### PR TITLE
make UidReverseComparator transitive, order null elements to right

### PR DESCRIPTION
**Pull request by Valodim**
this makes the comparator transitive by sorting null elements to the end. I'm not really sure about the assumptions made on the `uid` strings, so they might be better sorted to the start of the array? changing that should be trivial either way.

### CODE REVIEW

File: k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java [hunk 1]
```diff
@@ -137,22 +137,30 @@
     public static class UidReverseComparator implements Comparator<Message> {
         @Override
         public int compare(Message o1, Message o2) {
-            if (o1 == null || o2 == null || o1.getUid() == null || o2.getUid() == null) {
-                return 0;
-            }
-            int id1, id2;
+            Integer id1, id2;
```
Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java]: I'd rename `o1` and `o2` to `messageLeft` and `messageRight`. `id1` and `id2` to `uidLeft` and `uidRight`.

File: k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java [hunk 2]
```diff
@@ -137,22 +137,30 @@
...
+                id1 = null;
+            }
+            try {
                 id2 = Integer.parseInt(o2.getUid());
-            } catch (NumberFormatException e) {
+            } catch (NullPointerException | NumberFormatException e) {
+                id2 = null;
+            }
+            // noinspection NumberEquality, this checks if both are null
+            boolean bothAreNull = id1 == id2;
```
Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java]: This comparison could also be true if `id1` and `id2` are small integers where the same `Integer` instance is used from the integer cache. The method will still return the correct value, but the path taken is unexpected.

I propose to change the code to the following which is more explicit for all cases:

```java
if (id1 == null && id2 == null) {
    return 0;
} else if (id1 == null) {
    return 1;
} else if (id2 == null) {
    return -1;
}
```

File: k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java [hunk 3]
```diff
@@ -137,22 +137,30 @@
...
-                return -1;
-            return 0;
+            // check if either is null, and order the null one to the end
+            boolean eitherIsNull = (id1 == null) || (id2 == null);
+            if (eitherIsNull) {
+                return id1 == null ? 1 : -1;
+            }
+
+            // reverse order
+            return id2 -id1;
```
Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/controller/MessagingController.java]: I'd use `id2.compareTo(id1)` for better readability.

**Comment by cketti** [comment]
For tests please create one method per thing you test.
I also like to split tests in the phases *setup*, *exercise*, *verify*, e.g.

```java
@Test
public void compare_withTwoNullArguments_shouldReturnZero() throws Exception {
    Message messageLeft = null;
    Message messageRight = null;

    int result = comparator.compare(messageLeft, messageRight);

    assertEquals("Result must be 0 when both arguments are null", 0, result);
}
```

With this class I don't see the benefit of testing with random numbers. Just pick some constants and make sure the tests exercise all execution paths.
Also, only test the contract (`0`, `< 0`, or `> 0`) not the implementation detail (`uid2 - uid1`).

As a general rule: Whenever you find yourself implementing some logic in a test take a step back and think about whether this is really necessary. Chances are the test code will be at least as complex as the code under test. That increases the likelihood of introducing new bugs or making the same mistakes as the code under test.

**Comment by Valodim** [comment]
I like randomized testing like that because it has a chance to catch edge cases you missed, the transitive property here is a good example of that. I don't really have strong feelings about this though, if you don't want tests like that I'll remove it.

**Comment by cketti** [comment]
Randomized testing certainly has its uses. But it's hard. To be useful you need to record the input so an eventual failure can be reproduced fast. Also, the test depends on some (complex) code that is different from the code under test (in your case the Arrays.sort() implementation and the logic inside the test). So it's easy to introduce errors that are not immediately obvious.

In my opinion testing for the three required properties with constant input values is sufficient. Just make sure you do it for all classes of input values (message == null, message.getUid() == null, message.getUid() not being a parsable number, message.getUid() being a parsable number).

**Comment by cketti** [comment]
I made some changes. Feedback welcome.

**Comment by Valodim** [comment]
looked it over, looks good to me

**Comment by cketti** [comment]
Thanks
